File: app/Models/MusicTrack.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class MusicTrack extends Model
{
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'title',
        'artist',
        'album',
        'genre',
        'duration',
        'file_url',
        'cover_image',
        'release_date',
        'language',
        'country',
        'lyrics',
        'is_local',
        'is_featured',
        'play_count',
        'status',
        'uploaded_by',
        'metadata',
    ];

    protected $casts = [
        'release_date' => 'date',
        'is_local' => 'boolean',
        'is_featured' => 'boolean',
        'metadata' => 'array',
    ];

    public function uploader()
    {
        return $this->belongsTo(User::class, 'uploaded_by');
    }

    public function playlists()
    {
        return $this->belongsToMany(Playlist::class, 'playlist_tracks')->withTimestamps();
    }

    public function posts()
    {
        return $this->hasMany(Post::class, 'music_track_id');
    }

    public function threads()
    {
        return $this->hasMany(Thread::class, 'music_track_id');
    }

    public function userFavorites()
    {
        return $this->belongsToMany(User::class, 'user_favorite_tracks')->withTimestamps();
    }

    public function scopeLocal($query)
    {
        return $query->where('is_local', true);
    }

    public function scopeFeatured($query)
    {
        return $query->where('is_featured', true);
    }

    public function scopeByGenre($query, $genre)
    {
        return $query->where('genre', $genre);
    }

    public function scopePopular($query)
    {
        return $query->orderBy('play_count', 'desc');
    }

    public function getDurationFormattedAttribute()
    {
        $minutes = floor($this->duration / 60);
        $seconds = $this->duration % 60;
        return sprintf('%d:%02d', $minutes, $seconds);
    }
}

File: uniblend_mobile/uniblend_mobile/lib/shared/models/music_track.dart
class MusicTrack {
  final int id;
  final String title;
  final String artist;
  final String? album;
  final String? genre;
  final int? duration;
  final String fileUrl;
  final String? coverImage;
  final DateTime? releaseDate;
  final String language;
  final String country;
  final String? lyrics;
  final bool isLocal;
  final bool isFeatured;
  final int playCount;
  final String status;
  final int? uploadedBy;
  final Map<String, dynamic>? metadata;
  final DateTime createdAt;
  final DateTime updatedAt;

  MusicTrack({
    required this.id,
    required this.title,
    required this.artist,
    this.album,
    this.genre,
    this.duration,
    required this.fileUrl,
    this.coverImage,
    this.releaseDate,
    required this.language,
    required this.country,
    this.lyrics,
    required this.isLocal,
    required this.isFeatured,
    required this.playCount,
    required this.status,
    this.uploadedBy,
    this.metadata,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MusicTrack.fromJson(Map<String, dynamic> json) {
    return MusicTrack(
      id: json['id'],
      title: json['title'],
      artist: json['artist'],
      album: json['album'],
      genre: json['genre'],
      duration: json['duration'],
      fileUrl: json['file_url'],
      coverImage: json['cover_image'],
      releaseDate: json['release_date'] != null 
          ? DateTime.parse(json['release_date']) 
          : null,
      language: json['language'] ?? 'en',
      country: json['country'] ?? 'KE',
      lyrics: json['lyrics'],
      isLocal: json['is_local'] ?? true,
      isFeatured: json['is_featured'] ?? false,
      playCount: json['play_count'] ?? 0,
      status: json['status'] ?? 'active',
      uploadedBy: json['uploaded_by'],
      metadata: json['metadata'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'artist': artist,
      'album': album,
      'genre': genre,
      'duration': duration,
      'file_url': fileUrl,
      'cover_image': coverImage,
      'release_date': releaseDate?.toIso8601String(),
      'language': language,
      'country': country,
      'lyrics': lyrics,
      'is_local': isLocal,
      'is_featured': isFeatured,
      'play_count': playCount,
      'status': status,
      'uploaded_by': uploadedBy,
      'metadata': metadata,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  String get durationFormatted {
    if (duration == null) return '';
    final minutes = duration! ~/ 60;
    final seconds = duration! % 60;
    return '${minutes}:${seconds.toString().padLeft(2, '0')}';
  }

  String get flagEmoji {
    switch (country) {
      case 'KE':
        return '🇰🇪';
      case 'TZ':
        return '🇹🇿';
      case 'UG':
        return '🇺🇬';
      case 'RW':
        return '🇷🇼';
      default:
        return '🌍';
    }
  }

  String get genreDisplayName {
    switch (genre?.toLowerCase()) {
      case 'afrobeat':
        return 'Afrobeat';
      case 'benga':
        return 'Benga';
      case 'genge':
        return 'Genge';
      case 'gospel':
        return 'Gospel';
      case 'hip hop':
        return 'Hip Hop';
      case 'r&b':
        return 'R&B';
      case 'reggae':
        return 'Reggae';
      case 'traditional':
        return 'Traditional';
      default:
        return genre ?? 'Unknown';
    }
  }
}
